### PR TITLE
Get contract address from `CREATE` and `CREATE2` opcodes

### PR DESCRIPTION
**Pull request by tinom9**
### What was wrong?
Unlike `ethers.js` ([5.7.2](https://github.com/ethers-io/ethers.js/blob/v5.7.2/packages/address/src.ts/index.ts#L130)), `web3.py` does not include a utility method to deterministically determine the contract address from the inputs of `CREATE` and `CREATE2` opcodes.

Related to Issue #2762

### How was it fixed?
Added public utility methods to determine contract addresses.

### Todo:
- [X] Get contract address from `CREATE` opcode
- [X] Get contract address from `CREATE2` opcode
- [X] Documentation
- [X] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: web3/utils/address.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+
+def get_create_address(sender: HexAddress, nonce: Nonce) -> ChecksumAddress:
+    """
+    Determine the resulting `CREATE` opcode contract address for a sender and a nonce.
+    """
+    bytes_sender = bytes.fromhex(sender[2:])
+
+    contract_address = keccak(encode([bytes_sender, nonce])).hex()[-40:]
```
Comment (fselmo) [inline web3/utils/address.py:25]: nit: Ideally we import `rlp` and use `rlp.encode` here for clarification. We've done the same with `import abi` -> `abi.encode()` and it just makes it easier to read the code.

```suggestion
    contract_address = keccak(rlp.encode([bytes_sender, nonce])).hex()[-40:]
```

File: web3/utils/address.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def get_create2_address(
+    sender: HexAddress, salt: HexStr, init_code: HexStr
+) -> ChecksumAddress:
+    """
+    Determine the resulting `CREATE2` opcode contract address for a sender, salt and
+    bytecode.
+    """
+    bytes_prefix = bytes.fromhex("0xFF"[2:])
```
Comment (fselmo) [inline web3/utils/address.py:36]: nit: I could go either way here but `to_bytes` from the `eth-utils` dependency would not only make this bit of code more similar to the rest of the code base (so faster to read / process), but it also would accept both 0x-prefixed and non-0x-prefixed hexes.

Something like: 

```py
contract_address = keccak(
    b"\xff"
    + to_bytes(hexstr=sender)
    + to_bytes(hexstr=salt)
    + keccak(to_bytes(hexstr=init_code))
).hex()[-40:]
```

Thoughts on that?

File: web3/utils/address.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    HexStr,
+    Nonce,
+)
+
+
+def get_create_address(sender: HexAddress, nonce: Nonce) -> ChecksumAddress:
+    """
+    Determine the resulting `CREATE` opcode contract address for a sender and a nonce.
+    """
+    bytes_sender = bytes.fromhex(sender[2:])
```
Comment (fselmo) [inline web3/utils/address.py:23]: nit: See comment below, maybe use `to_bytes(hexstr=sender)` here?

**Review by fselmo — APPROVED** [review]
@tinom9 this is a great addition to the public utilities, thank you so much. We'd talked about it at some point and just hadn't gotten around to it. This is good to go as far as I'm concerned, I just had some nits I wanted to pick your brain on. Thoughts on those?

File: web3/utils/address.py (hunk 1 above)
Comment (tinom9) [reply web3/utils/address.py:25]: It's more readable as `rlp.encode`, agreed!

File: web3/utils/address.py (hunk 2 above)
Comment (tinom9) [reply web3/utils/address.py:36]: I slightly prefer `bytes.fromhex` because it's built-in, but code readability is king!

**Comment by fselmo** [comment]
Thanks again @tinom9!

**Pull request merged**